### PR TITLE
Add back off dead letter error handlers

### PR DESCRIPTION
**Pull request by jakekeeys**

### CODE REVIEW

File: backoff/backoff.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+func New(handler pubsub.ConsumerMessageHandler) ExponentialBackOffRetryingErrorHandler {
+	return NewWithFallBack(handler, func(msg pubsub.ConsumerMessage, err error) error {
+		return err
+	})
+}
+
+func NewWithFallBack(handler pubsub.ConsumerMessageHandler, errHandler pubsub.ConsumerErrorHandler) ExponentialBackOffRetryingErrorHandler {
+	return func(msg pubsub.ConsumerMessage, err error) error {
+		for i := 0; i < 3; i++ {
```
Comment (mjgarton) [inline backoff/backoff.go]: Does it make sense to let the caller specify the number of retries instead of hardcode 3?

Comment (jakekeeys) [reply backoff/backoff.go]: I think so

File: deadletter/deadletter.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+		messageTopic,
+	)
+}
+
+// NewWithFallBack returns a new DeadLetteringErrorHandler with fall back
+func NewWithFallBack(sink pubsub.MessageSink, errHandler pubsub.ConsumerErrorHandler, messageTopic string) DeadLetteringErrorHandler {
+	return func(msg pubsub.ConsumerMessage, err error) error {
+		failedMsg := FailedConsumerMessage{
+			Message:       msg,
+			MessageBase64: base64.StdEncoding.EncodeToString(msg.Data),
```
Comment (mjgarton) [inline deadletter/deadletter.go]: Why is this included as base64 as well as the []byte version?  Especially given that the json serialisation below will turn the bytes into base64 anyway.

Comment (mjgarton) [reply deadletter/deadletter.go]: (Also, putting the message in twice may significantly increase the risk of overflowing message size limits.)

Comment (jakekeeys) [reply deadletter/deadletter.go]:
```type TestR struct {
	Message []byte `json:"msg"`
}

func Test(t *testing.T) {
	test := TestR{}
	test.Message = []byte{123, 123, 123, 123, 123}
	JSON, _ := json.Marshal(test)
	log.Print(JSON)
	log.Print(base64.StdEncoding.EncodeToString(JSON))
}

2017/09/11 12:05:56 [123 34 109 115 103 34 58 34 101 51 116 55 101 51 115 61 34 125]

2017/09/11 12:05:56 eyJtc2ciOiJlM3Q3ZTNzPSJ9```

Comment (catalin-ilea) [reply deadletter/deadletter.go]: We're doing something similar in billing. Here's what our model looks like: https://github.com/utilitywarehouse/uw-lib-billing/blob/master/usage/err/error.go

Comment (mjgarton) [reply deadletter/deadletter.go]: @catalin-ilea Any parts of that you think should be adopted here?

Comment (catalin-ilea) [reply deadletter/deadletter.go]: Unless you plan on using one error topic per service it would be good to add the originating service to the payload. Also marking an error `Recoverable` or not was useful to us to distinguish business errors from blips with the infrastructure as you usually treat those differently.

Comment (mjgarton) [reply deadletter/deadletter.go]: Anything backwards incompatible that should stop this being merged not and iterated upon?
